src/psyclone/tests/nemo/transformations/openacc/data_directive_test.py

```diff
@@ -0,0 +1,536 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''Module containing py.test tests for the transformation of the PSy
+   representation of NEMO code using the OpenACC data directive.
+
+'''
+
+from __future__ import print_function, absolute_import
+import os
+import pytest
+from fparser.common.readfortran import FortranStringReader
+from psyclone.parse import parse
+from psyclone.psyGen import PSyFactory, TransInfo, InternalError
+from psyclone.transformations import TransformationError
+
+
+# Constants
+API = "nemo"
+# Location of the Fortran files associated with these tests
+BASE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                         "../../test_files")
+
+# Test code with explicit NEMO-style do loop
+EXPLICIT_DO = ("program explicit_do\n"
+               "  REAL, DIMENSION(jpi, jpj, jpk) :: umask\n"
+               "  DO jk = 1, jpk\n"
+               "     DO jj = 1, jpj\n"
+               "        DO ji = 1, jpi\n"
+               "           umask(ji,jj,jk) = ji*jj*jk/r\n"
+               "        END DO\n"
+               "     END DO\n"
+               "  END DO\n"
+               "end program explicit_do\n")
+
+
+def test_explicit(parser):
+    '''
+    Check code generation for enclosing a single explicit loop containing a
+    kernel inside a data region.
+
+    '''
+    reader = FortranStringReader(EXPLICIT_DO)
+    code = parser(reader)
+    psy = PSyFactory(API, distributed_memory=False).create(code)
+    schedule = psy.invokes.get('explicit_do').schedule
+    acc_trans = TransInfo().get_trans_name('ACCDataTrans')
+    schedule, _ = acc_trans.apply(schedule.children)
+    gen_code = str(psy.gen)
+
+    assert ("  REAL, DIMENSION(jpi, jpj, jpk) :: umask\n"
+            "  !$ACC DATA COPYOUT(umask)\n"
+            "  DO jk = 1, jpk") in gen_code
+
+    assert ("  END DO\n"
+            "  !$ACC END DATA\n"
+            "END PROGRAM explicit_do") in gen_code
+
+
+def test_data_no_gen_code():
+    ''' Check that the ACCDataDirective.gen_code() method raises the
+    expected error. '''
```

Comment:
Please explain what this expected error is.

Reply:
Done.